# Exploring Ebay Car Sales

In this project we'll work with a dataset of used cars from *eBay Kleinanzeigen*, a classifieds section of the German eBay website.

The original dataset was on Kaggle. Dataset provided by [Dataquest](https://www.dataquest.io/) is modified for the purposes of this project: 50,000 data points are sampled and the data set is dirtier to resemble more closely to a scraped data set.

The data dictionary provided with data is as follows:

- `dateCrawled` - when this ad was first crawled; all field-values are taken from this date
- `name` - name of the car
- `seller` - whether the seller is private or a dealer
- `offerType` - the type of listing
- `price` - the price on the ad to sell the car
- `abtest` - whether the listing is included in an A/B test
- `vehicleType` - the vehicle type
- `yearOfRegistration` - the year in which the car was first registered
- `gearbox` - the transmission type
- `powerPS` - the power of the car in PS
- `model` - the car model name
- `odometer` - how many kilometers the car has driven
- `monthOfRegistration` - the month in which the car was first registered
- `fuelType` - what type of fuel the car uses
- `brand` - the brand of the car
- `notRepairedDamage` - if the car has a damage which is not yet repaired
- `dateCreated` - the date on which the eBay listing was created
- `nrOfPictures` - the number of pictures in the ad
- `postalCode` - the postal code for the location of the vehicle
- `lastSeenOnline` - when the crawler saw this ad last online

The aim of this project is to clean the data and analyze used car listings. 

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

### Cleaning column names

The dataset contains 20 columns. Columns with integers are `yearOfRegistration`, `powerPS`, `monthOfRegistration`, `nrOfPictures` and `postalCode`, but the majority of columns contains strings. There are columns with null values: `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage`, but none with more than 20% of null values. Column names are in camelcase format.

We will convert column names to snakecase format and reword some of the column names based on the data dictionary to be more descriptive:
   - `earOfRegistration` to `registration_year`
   - `monthOfRegistration` to `registration_month`
   - `notRepairedDamage` to `unrepaired_damage`
   - `dateCreated` to `ad_created`

Column names:

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
column_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.columns = column_names
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Initial exploration and cleaning

We can check for text columns where all or almost all values are the same (these can be dropped as they don't have useful information for analysis) and numeric data stored as text which can be cleaned and converted.

In [5]:
autos.shape

(50000, 20)

In [6]:
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

We can see that:
- columns `seller` and `offer_type` have mostly one value
- columns `price` and `odometer` are numeric values stored as text that need to be cleaned

In [8]:
autos["price"].value_counts().head()

$0        1421
$500       781
$1,500     734
$2,500     643
$1,000     639
Name: price, dtype: int64

In [9]:
autos["price"]= autos["price"].str.replace("$", "", regex=False)
autos["price"]= autos["price"].str.replace(",", "", regex=False)
autos["price"]= autos["price"].astype(int)

autos["price"].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [10]:
autos["odometer"].value_counts().head()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
Name: odometer, dtype: int64

In [11]:
autos["odometer"]= autos["odometer"].str.replace("km", "", regex=False)
autos["odometer"]= autos["odometer"].str.replace(",", "", regex=False)
autos["odometer"]= autos["odometer"].astype(int)

autos["odometer"].value_counts().head()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer, dtype: int64

In [12]:
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

### Exploring odometer and price columns

To check column `odometer_km`:

In [13]:
autos["odometer_km"].value_counts().count()

13

In [14]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [15]:
autos["odometer_km"].value_counts().sort_index(ascending= False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In the `odometer_km` column we have no specific outliers, unexpected values and/or scarce frequencies.

To check the `price` column:

In [16]:
autos["price"].value_counts().count()

2357

In [17]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [18]:
autos["price"].value_counts().sort_index(ascending= True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

In [19]:
autos["price"].value_counts().sort_index(ascending= False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

Looking at the unique values of the `price` column, in the lowest 20 values (0-35 USD) there are:
   - for 0 USD: 1421 entries
   - for 1 USD: 156 entries
   - for 2-35 USD: under 10 entries each (mostly under 5 entries)

Since the price 0 USD or 1 USD can also mean an intention to give or trade the car, and entries from 2-35 USD seem consistent, we will keep this data.

In the highest values, there are some odd entries. Let's take a look at entries above 350,000 USD:

In [20]:
autos["price"].loc[lambda x : x>350000].value_counts().sort_index(ascending= False)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

The highest entry is 99,999,999 USD which is most probably an unrealistic price for a car in classifieds. We can also relatively safely extend this to 1,300,000 USD. Even if we consider 1 million USD to be a realistic maximum price, the entries as 1,234,567 USD, 999,999 USD  and 999,990 USD don't seem plausible enough. So we will remove highest 10 price points, ranging from 99,999,999 - 999,990 USD, or in total 14 rows/data entries (0.028% of data).

In [21]:
autos_backup_copy = autos
autos_clean = autos

In [22]:
autos_clean = autos[autos["price"].between(0,350000)]

In [23]:
autos_clean.shape

(49986, 20)

In [24]:
autos_clean["price"].value_counts().count()

2347

In [25]:
autos_clean["price"].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64

The outliers are removed. The mean of the `price` column is approximately half of the previous value and standard deviation is much smaller.

### Exploring the date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

`date_crawled`, `last_seen` and `ad_created` columns are all strings:

In [26]:
autos_clean[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

These columns all represent full timestamp values, where the first 10 characters represent the day. We will extract the date and analyze the data.

In [27]:
autos_clean['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head(20)

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
Name: date_crawled, dtype: float64

The distribution of values is relatively uniform, starting with the date 2016-03-05. This is expected, as the crawling happens at regular intervals to catch all new ads that showed up.

In [28]:
autos_clean['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head(20)

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002
2016-01-10    0.00004
2016-01-13    0.00002
2016-01-14    0.00002
2016-01-16    0.00002
2016-01-22    0.00002
2016-01-27    0.00006
2016-01-29    0.00002
2016-02-01    0.00002
2016-02-02    0.00004
2016-02-05    0.00004
2016-02-07    0.00002
2016-02-08    0.00002
Name: ad_created, dtype: float64

The distribution of values is relatively uniform, starting with the date 2015-06-11, but the percentages are very low, going from 0.002% upward. This is expected, as the ads are places individually by people every day.

In [29]:
autos_clean['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head(20)

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
Name: last_seen, dtype: float64

The distribution of values is more diverse and starting with the date 2016-03-05. It shows the user browsing trend.

In [30]:
autos_clean['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Registration year data

In the `registration_year` column minimum is 1000 and maximum 9999. No car could be registered in the middle ages or in the future. A car can't also be registered for the first time after the listing was seen, so registration year above 2016 is inaccurate. Earliest valid registration year is probably in the 1900s.

In [31]:
autos_clean['registration_year'].value_counts().sort_index(ascending=False).head(20)

9999       4
9996       1
9000       2
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       3
2018     491
2017    1452
2016    1316
2015     399
2014     665
2013     806
2012    1323
2011    1634
Name: registration_year, dtype: int64

In [32]:
autos_clean['registration_year'].value_counts().sort_index(ascending=True).head(20)

1000    1
1001    1
1111    1
1500    1
1800    2
1910    9
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    3
1951    2
1952    1
1953    1
Name: registration_year, dtype: int64

Rows we can remove are `registration_year` values from 2017 to 9999 and from 1000 to 1800. This means we should leave values from 1910 to 2016:

In [33]:
autos_clean = autos_clean[autos_clean['registration_year'].between(1910,2016)]

In [34]:
autos_clean.shape

(48016, 20)

In [35]:
autos_clean['registration_year'].describe()

count    48016.000000
mean      2002.806002
std          7.306212
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [36]:
autos_clean['registration_year'].value_counts(normalize=True).sort_index(ascending=False).head(20)

2016    0.027408
2015    0.008310
2014    0.013850
2013    0.016786
2012    0.027553
2011    0.034030
2010    0.033260
2009    0.043673
2008    0.046464
2007    0.047984
2006    0.056377
2005    0.062792
2004    0.057002
2003    0.056794
2002    0.052753
2001    0.056273
2000    0.069852
1999    0.062438
1998    0.051087
1997    0.042236
Name: registration_year, dtype: float64

Registration year 2016 holds 2.74% of values in the distribution and 2015 only 0.83%. Years 2006 to 1998 have over 5% and year 2000 has 6.99%.

In [37]:
autos_clean['registration_year'].value_counts(normalize=True).sort_index(ascending=True).head(20)

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
Name: registration_year, dtype: float64

From 1910 to 1958 the percentage is significantly lower, in the range 0.002-0.019%.

### Exploring price by brand

Let's take a look at the `brand` column:

In [38]:
autos_clean['brand'].describe()

count          48016
unique            40
top       volkswagen
freq           10185
Name: brand, dtype: object

In [39]:
autos_clean['brand'].value_counts().head(40)

volkswagen        10185
bmw                5283
opel               5194
mercedes_benz      4579
audi               4149
ford               3350
renault            2274
peugeot            1418
fiat               1242
seat                873
skoda               770
mazda               727
nissan              725
smart               668
citroen             668
toyota              599
sonstige_autos      523
hyundai             473
volvo               444
mini                415
mitsubishi          391
honda               377
kia                 341
alfa_romeo          318
porsche             293
suzuki              284
chevrolet           274
chrysler            176
daihatsu            123
dacia               123
jeep                108
subaru              105
land_rover           98
saab                 77
jaguar               76
trabant              75
daewoo               72
rover                65
lancia               52
lada                 29
Name: brand, dtype: int64

There are 40 unique values in the `brand` column. To find brands that have over 5% of the total number of cars, number of entries for the brand should be over:

In [40]:
round(len(autos_clean['brand'])*0.05)

2401

The brands that have more than 2401 entries are:

In [41]:
autos_clean['brand'].value_counts().loc[lambda x: x>2400]

volkswagen       10185
bmw               5283
opel              5194
mercedes_benz     4579
audi              4149
ford              3350
Name: brand, dtype: int64

In [42]:
mp = autos_clean.groupby('brand')['price'].mean().filter(items=['volkswagen', 'opel', 'bmw', 'mercedes_benz','audi','ford']).sort_values(ascending=False)
mp

brand
audi             9093.650036
mercedes_benz    8485.239572
bmw              8102.536248
volkswagen       5231.081983
ford             3652.095224
opel             2876.716404
Name: price, dtype: float64

The highest mean value of the price in top 5% holds `audi`, followed by `mercedes_benz` and `bmw`.

The lowest mean value of the price in top 5% holds `opel`, followed by `ford` and`volkswagen`.

### Average mileage

To calculate the mean mileage for 6 top brands and store it in a dictionary:

In [43]:
mm = autos_clean.groupby('brand')['odometer_km'].mean().filter(items=['volkswagen', 'opel', 'bmw', 'mercedes_benz','audi','ford']).sort_values(ascending=False)
mm

brand
bmw              132431.383684
mercedes_benz    130856.082114
audi             129287.780188
opel             129223.142087
volkswagen       128724.104075
ford             124068.656716
Name: odometer_km, dtype: float64

To combine price and mileage data:

In [44]:
df = pd.DataFrame()
df['mean_price'] = mp
df['mean_mileage'] = mm
df

mean_price   mean_mileage
brand                                    
audi           9093.650036  129287.780188
mercedes_benz  8485.239572  130856.082114
bmw            8102.536248  132431.383684
volkswagen     5231.081983  128724.104075
ford           3652.095224  124068.656716
opel           2876.716404  129223.142087

There is no direct link between the mean price and mean mileage for top 6 brands.